visualization/app/codeCharta/ui/screenshotButton/screenshotButton.component.spec.ts

```diff
@@ -1,95 +1,39 @@
+import { ThreeCameraServiceToken, ThreeRendererServiceToken, ThreeSceneServiceToken } from "../../services/ajs-upgraded-providers"
+import userEvent from "@testing-library/user-event"
+
+describe("screenshotButtonComponent", () => {
 	beforeEach(() => {
-		restartSystem()
-		mockLoadScript()
-		rebuildController()
-	})
-
-	function restartSystem() {
-		instantiateModule("app.codeCharta.ui.screenshotButton")
-
-		storeService = getService<StoreService>("storeService")
-		$rootScope = getService<IRootScopeService>("$rootScope")
-		threeSceneService = getService<ThreeSceneService>("storeService")
-		threeCameraService = getService<ThreeCameraService>("storeService")
-		threeRendererService = getService<ThreeRendererService>("storeService")
-	}
-
-	function rebuildController() {
-		screenshotButtonController = new ScreenshotButtonController(
-			threeSceneService,
-			threeCameraService,
-			threeRendererService,
-			storeService,
-			$rootScope
-		)
-	}
-
-	describe("makeScreenshot", () => {
-		it("should call loadScript", () => {
-			screenshotButtonController["loadScript"] = jest.fn()
-			screenshotButtonController.makeScreenshotToFile()
-			expect(screenshotButtonController["loadScript"]).toBeCalled()
-		})
-		it("should call makePNGFileName", () => {
-			screenshotButtonController["makePNGFileName"] = jest.fn()
-			screenshotButtonController.makeScreenshotToFile()
-			expect(screenshotButtonController["makePNGFileName"]).toBeCalled()
-		})
-
-		it("should call buildScreenShotCanvas", () => {
-			screenshotButtonController["buildScreenShotCanvas"] = jest.fn()
-			screenshotButtonController.makeScreenshotToFile()
-			expect(screenshotButtonController["buildScreenShotCanvas"]).toBeCalled()
+		TestBed.configureTestingModule({
+			imports: [ScreenshotButtonModule],
+			providers: [
+				{ provide: ThreeCameraServiceToken, useValue: {} },
+				{ provide: ThreeSceneServiceToken, useValue: {} },
+				{ provide: ThreeRendererServiceToken, useValue: {} }
+			]
 		})
 	})
 
-	describe("onScreenshotToClipboardEnabledChanged", () => {
-		it("should set screenshotToClipboardEnabled in viewModel", () => {
-			storeService.dispatch(setScreenshotToClipboardEnabled(true))
+	it("should copy to clipboard on click, when screenshot to clipboard is enabled", async () => {
+		Store.store.dispatch(setScreenshotToClipboardEnabled(true))
+		const { fixture } = await render(ScreenshotButtonComponent, { excludeComponentDeclaration: true })
+		fixture.componentInstance.makeScreenshotToClipBoard = jest.fn()
 
-			screenshotButtonController.onScreenshotToClipboardEnabledChanged(
-				storeService.getState().appSettings.screenshotToClipboardEnabled
-			)
-
-			expect(screenshotButtonController["_viewModel"].screenshotToClipboardEnabled).toBe(true)
-		})
+		userEvent.click(screen.getByTitle("Copy screenshot to clipboard (Ctrl+Alt+F), export it as a file by (Ctrl+Alt+S)"))
+		expect(fixture.componentInstance.makeScreenshotToClipBoard).toHaveBeenCalled()
```

Comment:
This test would pass if we replaced
`
makeScreenShotToClipBoard = ( ) => { }`

Is there a way how we can make this test more "demanding" (=stricter assert part)?

Reply:
I like your way of thinking! 👍 

I applied the boys scout rule here. In an ideal world, the UI shouldn't contain the logic behind it at all, but just use some service, which does all the calculation. Then the UI has only the current test to use the service and the logic should be tested within the service.

Unfortunately the logic was already in the component and it wasn't tested but mocked away at a deeper level. 

So the tests cover the same as before and now we have an Angular component more and an AngularJS component less. Small improvements 🤓

Reply:
I added a reminder to #2410. Maybe we can add a nice e2e test later. But lets keep with small improvements (Angular Migration) for now. Although 👍  for the spirit!